### PR TITLE
8292446: Make TableRateStatistics optional in CHT

### PR DESCRIPTION
**Pull request by jdksjolen**
Right now the footprint of disabling statistics for tables where they're not used is minimal as they're single instance and static. We're working on using CHT in more contexts, where it will be dynamically allocated, and then these changes will add up to something more substantial.

Passes tier1+tier2

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292446](https://bugs.openjdk.org/browse/JDK-8292446): Make TableRateStatistics optional in CHT

### Reviewers
 * [Coleen Phillimore](https://openjdk.org/census#coleenp) (@coleenp - **Reviewer**)
 * [Robbin Ehn](https://openjdk.org/census#rehn) (@robehn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9899/head:pull/9899` \
`$ git checkout pull/9899`

Update a local copy of the PR: \
`$ git checkout pull/9899` \
`$ git pull https://git.openjdk.org/jdk pull/9899/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9899`

View PR using the GUI difftool: \
`$ git pr show -t 9899`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9899.diff">https://git.openjdk.org/jdk/pull/9899.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/utilities/concurrentHashTable.inline.hpp [hunk 1]
```diff
@@ -1028,6 +1030,7 @@ inline ConcurrentHashTable<CONFIG, F>::
   delete _resize_lock;
   free_nodes();
   delete _table;
+  delete _stats_rate;
```
Comment (zhengyu123) [inline src/hotspot/share/utilities/concurrentHashTable.inline.hpp:1033]: `_stats_rate` might not be initialized if `enable_statistics == false` in constructor, could crash JVM here.

File: src/hotspot/share/gc/g1/g1CardSet.cpp [hunk 2]
```diff
@@ -34,6 +34,7 @@
 #include "runtime/globals_extension.hpp"
 #include "runtime/java.hpp"
 #include "utilities/bitMap.inline.hpp"
+#include "utilities/concurrentHashTable.hpp"
```
Comment (coleenp) [inline src/hotspot/share/gc/g1/g1CardSet.cpp:37]: Do you needs this? Doesn't concurrentHashTable.inline.hpp include it already?

**Review by coleenp — COMMENTED** [review]
Thank you for doing this. I have a couple small comments.

File: src/hotspot/share/utilities/concurrentHashTable.hpp [hunk 3]
```diff
@@ -42,6 +42,22 @@ template <typename CONFIG, MEMFLAGS F>
 class ConcurrentHashTable : public CHeapObj<F> {
   typedef typename CONFIG::Value VALUE;
  private:
+  // _stats_rate is null if statistics are not enabled.
+  TableRateStatistics* _stats_rate  = nullptr;
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/utilities/concurrentHashTable.hpp:46]: My Google-fu fails me in what the difference between this assignment and doing it in the constructor is, so I'm moving it to the constructor to be sure.

File: src/hotspot/share/utilities/concurrentHashTable.hpp [hunk 4]
```diff
@@ -376,18 +387,22 @@ class ConcurrentHashTable : public CHeapObj<F> {
...
+  static const size_t DEFAULT_MAX_SIZE_LOG2 = 21;
+  static const size_t DEFAULT_START_SIZE_LOG2 = 13;
+  static const size_t DEFAULT_GROW_HINT = 4; // Chain length
+  static const bool DEFAULT_ENABLE_STATISTICS = true;
   ConcurrentHashTable(size_t log2size = DEFAULT_START_SIZE_LOG2,
                       size_t log2size_limit = DEFAULT_MAX_SIZE_LOG2,
                       size_t grow_hint = DEFAULT_GROW_HINT,
-                      void* context = NULL);
+                      void* context = nullptr,
+                      bool enable_statistics = DEFAULT_ENABLE_STATISTICS);
```
[the remark this answers is not in the document]
Comment (jdksjolen) [reply src/hotspot/share/utilities/concurrentHashTable.hpp]: That did make things nicer!

File: src/hotspot/share/gc/g1/g1CardSet.cpp (hunk 2 above)
Comment (jdksjolen) [reply src/hotspot/share/gc/g1/g1CardSet.cpp]: Fixed :).

File: src/hotspot/share/utilities/concurrentHashTable.hpp [hunk 5]
```diff
@@ -376,18 +387,22 @@ class ConcurrentHashTable : public CHeapObj<F> {
   void delete_in_bucket(Thread* thread, Bucket* bucket, LOOKUP_FUNC& lookup_f);
 
  public:
+  // Default sizes
+  static const size_t DEFAULT_MAX_SIZE_LOG2 = 21;
+  static const size_t DEFAULT_START_SIZE_LOG2 = 13;
+  static const size_t DEFAULT_GROW_HINT = 4; // Chain length
+  static const bool DEFAULT_ENABLE_STATISTICS = true;
```
Comment (coleenp) [inline src/hotspot/share/utilities/concurrentHashTable.hpp:394]: Ok, one last thing. I wonder if we should default to false for enabling table rate statistics?  Only the StringTable and SymbolTable need it.

Comment (jdksjolen) [reply src/hotspot/share/utilities/concurrentHashTable.hpp]: Fixed.

**Review by coleenp — APPROVED** [review]
This looks really good! Thank you for fixing this!

**Review by robehn — APPROVED** [review]
Thanks for fixing!

**Pull request closed without merging**